### DISCUSSION TITLE
data leakage

### DISCUSSION THREAD

**Discussion by bbb801** [blue-yonder/tsfresh#1084] [Q&A]
Dear Sir/Madam,

I noticed in your '01 Feature Extraction and Selection.ipynb' that it first uses feature extraction and then splits the data into train and test. Is it leading to data leakage? Can we split the data first and then train the model to do feature extraction on the test set?

**Comment by nils-braun** [comment]
Hi @bbb801 - yes, you are indeed correct. This is not the best practice and therefore, our example should not include it.
Would you like to do a PR and change this? Thank you :)

**Reply by bbb801** [reply]
Thank you. I have read your notebooks and don't know how to apply it to test set.

**Reply by nils-braun** [reply]
Hi! I think you do not need to deal with the test set in the notebook. If you just split the initial data into train and test set and only apply the feature extraction and the rest to the train set, that is also fine. 
Another possibility would be mentioning that you only deal with the train set in this notebook at all and do not do any splitting (also not for the training of the classifier)

**Reply by bbb801** [reply]
Thank you. I will be applying this tool to my own data, and I have already split the data into training and test sets. I am now considering how to apply it to the test set. The only way I can think of is to use extract_features separately on the training and test sets. Is there another way to do this?

**Reply by bbb801** [reply]
However, when I read your '02 sklearn Pipeline.ipynb', it seemed statistics from training can be applied to the test set. May I know these steps still lead to data leakage? I am wondering if 'ppl.set_params(augmenter__timeseries_container=df_ts);' will compute the statistics from all data (including train and test set), and thus lead to data leakage. Thank you.

![image](https://github.com/user-attachments/assets/f578132a-d697-4633-a2f2-68c8ebb5471c)

**Reply by nils-braun** [reply]
No, there is no data leakage. The example in the docs gibes some more information: https://tsfresh.readthedocs.io/en/latest/api/tsfresh.transformers.html#tsfresh.transformers.relevant_feature_augmenter.RelevantFeatureAugmenter
You fit it on the train set but you predict on the test set.

**Reply by nils-braun** [reply]
Would you still like to do a PR to fix the example notebook?

**Reply by bbb801** [reply]
Thank you for your explanation. I checked your codes and found that I might not be able to fix them. I am just a beginner. Sorry.
By the way could you tell me where does 'read_in_timeseries()' come from (https://tsfresh.readthedocs.io/en/latest/api/tsfresh.transformers.html#tsfresh.transformers.relevant_feature_augmenter.RelevantFeatureAugmenter) ? I checked all of your codes and cannot import it.

![image](https://github.com/user-attachments/assets/8db16e5b-7b37-4778-be37-ea4db90c46ef)

**Reply by nils-braun** [reply]
The `read_in_timeseries` is the function you need to write. This is where you load your own data, so we can not provide this function for you :)
